tools/clang/unittests/HLSL/ExecutionTest.cpp

```diff
@@ -9740,99 +9756,130 @@ TEST_F(ExecutionTest, DynamicResourcesUniformIndexingTest) {
+        continue;
+      }
+      D3D12_FEATURE_DATA_D3D12_OPTIONS devOptions;
+      VERIFY_SUCCEEDED(
+          pDevice->CheckFeatureSupport((D3D12_FEATURE)D3D12_FEATURE_D3D12_OPTIONS,
+                                       &devOptions, sizeof(devOptions)));
+      if (devOptions.ResourceBindingTier < D3D12_RESOURCE_BINDING_TIER_3) {
+        WEX::Logging::Log::Comment(
+            L"Device does not support Resource Binding Tier 3");
+        WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+        return;
+      }
 
-    CComPtr<ID3D12Device> pDevice;
-    if (!CreateDevice(&pDevice, sm, false /* skipUnsupported */)) {
-      continue;
-    }
-    D3D12_FEATURE_DATA_D3D12_OPTIONS devOptions;
-    VERIFY_SUCCEEDED(
-        pDevice->CheckFeatureSupport((D3D12_FEATURE)D3D12_FEATURE_D3D12_OPTIONS,
-                                     &devOptions, sizeof(devOptions)));
-    if (devOptions.ResourceBindingTier < D3D12_RESOURCE_BINDING_TIER_3) {
-      WEX::Logging::Log::Comment(
-          L"Device does not support Resource Binding Tier 3");
-      WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
-      return;
-    }
+      // Add compile options
+      char compilerOptions[256];
+      if (non_uniform_bit && sm==D3D_SHADER_MODEL_6_0) {
+        sprintf_s(compilerOptions, sizeof(compilerOptions),
+                                "-D FALLBACK=%d, -D NON_UNIFORM=%d", 1, 1);
+      }
+      else if (non_uniform_bit || sm==D3D_SHADER_MODEL_6_0){
+        if (non_uniform_bit){
+          sprintf_s(compilerOptions, sizeof(compilerOptions),
+                                "-D NON_UNIFORM=%d",1);
+        } 
+        else{
+          sprintf_s(compilerOptions, sizeof(compilerOptions),
+                                "-D FALLBACK=%d",1);
+        }
+      }
 
-    // Test Compute shader
-    {
-      pShaderOp->CS = pShaderOp->GetString("CS66");
-      std::shared_ptr<ShaderOpTestResult> test = RunShaderOpTestAfterParse(
-          pDevice, m_support, "DynamicResourcesUniformIndexing", nullptr,
-          ShaderOpSet);
+      // by default a root value is added.
+      // remove the root value if this is the non-fallback path
+      if (sm==D3D_SHADER_MODEL_6_6 && non_uniform_bit)
+      {
+        pShaderOp->RootValues.clear();
+      }
 
-      MappedData resultData;
-      test->Test->GetReadBackData("g_result", &resultData);
-      const float *resultFloats = (float *)resultData.data();
+      // Update shader target in xml.
+      for (st::ShaderOpShader &S : pShaderOp->Shaders){
+        if (sm==D3D_SHADER_MODEL_6_0){
+          S.Arguments = compilerOptions;
+          vector<char> s;
+          for (unsigned int j = 0; j < sizeof(S.Target); j++){
+            s.push_back(S.Target[j]);
+          }
+          // change Target from ps_6_6 to ps_6_0, for example
+          s[sizeof(S.Target)-1] = '0';
+          strcpy_s(s.data(), sizeof(S.Target), S.Target);
```

Comment:
It's not good to modify the strings owned by ShaderOp in-place.  These are not expected to be mutable.  Plus sizeof() isn't going to give you the string length, it gives you the size of the type (LPCSTR), which for 64-bit architectures is 8 bytes.  You're thinking of `strlen()`, not `sizeof()`.

You can use this instead:
```
std::string Target(S.Target);
Target[Target.length() - 1] = '0';
S.Target = pShaderOp->GetString(Target.c_str());
```